pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -490,11 +490,13 @@ func (s *DiscoveryServer) WorkloadUpdate(id string, labels map[string]string, _
 	}
 	w, f := s.WorkloadsByID[id]
 	if !f {
-		// First time this workload has been seen. Likely never connected, no need to
-		// push
+		// First time this workload has been seen
 		s.WorkloadsByID[id] = &Workload{
 			Labels: labels,
 		}
+		// Although newly added, we are still pushing to this proxy to cover
+		// cases when it connected before getting a workload update event (#10904)
+		s.ConfigUpdate(true)
```

Comment:
how does this fix the issue of labels not being present on the service instances?

Reply:
In my experiments it wasn't an issue of labels not being present but 0 service instances.
It happens when new instance is XDS connecting before the k8s controller got an update event about its addition.

Reply:
The effect is the same isn't it? The gateway code does GetProxyServiceInstances() which will return empty set of instances if the pod does not have any associated instance. Triggering config Update wont help. Unfortunately, the ads logic thinks that anytime a service instance update event happens, it can be only for EDS, and not for LDS/CDS/RDS. and if you trigger a service instance update event for every service instance update, then you disable all the incremental EDS that @costinm put in. 

A temporary solution might be to add a delay in connecting to Pilot so that this race condition can be avoided. alternatively, there is code somewhere in ads.go or discovery.go that computes the service instances of the proxy connecting to Pilot. If the list of service instances is empty and its a gateway, you terminate the connection and force the proxy to reconnect..

A more reliable solution is to use a fallback logic in computing the labels of the instance in the gateway. If there are no proxy service instances, then use the labels in the workload's metadata field (I believe in pilot agent or helm chart, we attach the workload labels using ISTIO_META_..).